srml/contract/src/wasm/mod.rs

```diff
@@ -17,154 +17,166 @@
-	fn set_storage(&mut self, key: &[u8], value: Option<Vec<u8>>);
-
-	/// Create a new account for a contract.
+pub use self::code::save as save_code;
+
+/// A prepared wasm module ready for execution.
+#[derive(Clone, Encode, Decode)]
+pub struct PrefabWasmModule {
+	/// Version of the schedule with which the code was instrumented.
+	#[codec(compact)]
+	schedule_version: u32,
+	#[codec(compact)]
+	initial: u32,
+	#[codec(compact)]
+	maximum: u32,
+	/// This field is reserved for future evolution of format.
 	///
-	/// The newly created account will be associated with the `code`. `value` specifies the amount of value
-	/// transfered from this to the newly created account.
-	fn create(
-		&mut self,
-		code: &[u8],
-		value: BalanceOf<Self::T>,
-		gas_meter: &mut GasMeter<Self::T>,
-		data: &[u8],
-	) -> Result<CreateReceipt<Self::T>, ()>;
-
-	/// Call (possibly transfering some amount of funds) into the specified account.
-	fn call(
-		&mut self,
-		to: &AccountIdOf<Self::T>,
-		value: BalanceOf<Self::T>,
-		gas_meter: &mut GasMeter<Self::T>,
-		data: &[u8],
-		output_data: &mut Vec<u8>,
-	) -> Result<(), ()>;
-
-	/// Returns a reference to the account id of the caller.
-	fn caller(&self) -> &AccountIdOf<Self::T>;
+	/// Basically, for now this field will be serialized as `None`. In the future
+	/// we would be able to extend this structure with.
+	_reserved: Option<()>,
+	/// Code instrumented with the latest schedule.
+	code: Vec<u8>,
 }
 
-/// Error that can occur while preparing or executing wasm smart-contract.
-#[derive(Debug, PartialEq, Eq)]
-pub enum Error {
-	/// Error happened while serializing the module.
-	Serialization,
-
-	/// Error happened while deserializing the module.
-	Deserialization,
-
-	/// Internal memory declaration has been found in the module.
-	InternalMemoryDeclared,
+pub struct WasmExecutable {
+	entrypoint_name: &'static [u8],
+	prefab_module: PrefabWasmModule,
+}
 
-	/// Gas instrumentation failed.
-	///
-	/// This most likely indicates the module isn't valid.
-	GasInstrumentation,
+pub struct WasmLoader<'a, T: Trait> {
+	schedule: &'a Schedule<T::Gas>,
+}
 
-	/// Stack instrumentation failed.
-	///
-	/// This  most likely indicates the module isn't valid.
-	StackHeightInstrumentation,
+impl<'a, T: Trait> WasmLoader<'a, T> {
+	pub fn new(schedule: &'a Schedule<T::Gas>) -> Self {
+		WasmLoader { schedule }
+	}
+}
 
-	/// Error happened during invocation of the contract's entrypoint.
-	///
-	/// Most likely because of trap.
-	Invoke,
+impl<'a, T: Trait> ::exec::Loader<T> for WasmLoader<'a, T> {
+	type Executable = WasmExecutable;
 
-	/// Error happened during instantiation.
-	///
-	/// This might indicate that `start` function trapped, or module isn't
-	/// instantiable and/or unlinkable.
-	Instantiate,
+	fn load_init(&self, code_hash: &CodeHash<T>) -> Result<WasmExecutable, &'static str> {
+		let prefab_module = code::load::<T>(code_hash, self.schedule)?;
```

Comment:
Could be rewritten as:

`code::load::<T>(code_hash, self.schedule).map(|prefab_module|  WasmExecutable { entrypoint_name: b"deploy", prefab_module })`

Reply:
I have a slight preference avoiding combinators. But I think this one will work either way so I will happily merge your [gh-]suggestion if you send me one!